### DISCUSSION TITLE
Entity factories improvements

### DISCUSSION THREAD

**Discussion by samuelbednarcik** [mikro-orm/mikro-orm#3636] [Ideas]
**Is your feature request related to a problem? Please describe.**
I would like to implement more features to entity factories which helps a lot when writing a test or generating example data. My proposal is heavily inspired by [Laravel factories](https://laravel.com/docs/9.x/eloquent-factories) where these concepts are implemented and battle-tested. In case this makes sense to you, I would like to implement it.

**Describe the solution you'd like**

_Relationships as part of definition_
I would like to define relationships directly in the definition. It's annoying to manually fill them every time if I just don't care.

Example:
```
return {
  description: faker.lorem.paragraphs(2),
  author: UserFactory
};
```

This would require to create some global factory which discovers all factories ( something like migrations discovery? ) a then uses them to fill the field.

_States_
In case I have an entity with a lot of fields, it becomes hard to create an entity in a specific state. For example, I have an Auction and it can have many states ( active, waiting, closed, cancelled etc. ). Each state means different combination of columns with specific values.

Maybe we can add states method to factories?

Example:
```
states(): Record<string, EntityData<Auction>> {
  return {
    cancelled: {
      cancelledAt: new Date(),
    },
    active: {
      endsAt: new Date(), // generate date before now
    },
  };
}
```

**Describe alternatives you've considered**
I worked around these issues by writing my own helpers, but it would feel nice if they were implemented directly in MikroORM.

**Comment by B4nan** [comment]
Sure, sounds good to me!

**Comment by B4nan** [comment]
Any updates? I was expecting to see some PRs (as you said you even got it working locally).

edit: Moving to discussions, as said above, will be happy to accept a PR, but I don't plan to do any improvements myself.

**Reply by samuelbednarcik** [reply]
Don't have much time now, but I definitely plan to create PR for it.

**Comment by nexfader** [comment]
It can be solved right now using the `each` method:
```typescript
class AuctionFactory extends Factory<Auction> {
  model = Auction;

  definition() {
    return {
      description: faker.lorem.paragraphs(2),
      author: UserFactory,
      endsAt: new Date(), // Assuming the auction is active by default
    };
  }

  cancelled() {
    return this.each((auction) =>
      this.em.assign(auction, {
        cancelledAt: new Date(),
      }),
    );
  }
}
```

Usage:
```typescript
const auction = await new AuctionFactory(this.em).cancelled().createOne();
// Or multiple
const auctions = await new AuctionFactory(this.em).cancelled().create(5);
```
